src/mac/beacon.rs

```diff
@@ -0,0 +1,627 @@
+        let el = self.extended_address_count;
+        buf[0] = (((el as u8) << 4) & EXTENDED_MASK) | ((sl as u8) & SHORT_MASK);
+        let mut piece = &mut buf[1..];
+        for n in 0..self.short_address_count {
+            let addr = self.short_addresses[n];
+            addr.encode(piece);
+            piece = &mut piece[ss..];
+        }
+        for n in 0..self.extended_address_count {
+            let addr = self.extended_addresses[n];
+            addr.encode(piece);
+            piece = &mut piece[es..];
+        }
+        1 + sl * ss + sl * es
+    }
+    /// Get the short addresses
+    pub fn short_addresses(&self) -> &[ShortAddress] {
+        &self.short_addresses[..self.short_address_count]
+    }
+    /// Get the extended address
+    pub fn extended_addresses(&self) -> &[ExtendedAddress] {
+        &self.extended_addresses[..self.extended_address_count]
+    }
+}
+ /// Beacon frame
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct Beacon {
+    /// Superframe specification
+    pub superframe_spec: SuperframeSpecification,
+    /// Guaranteed time slot information
+    pub guaranteed_time_slot_info: GuaranteedTimeSlotInformation,
+    /// Pending addresses
+    pub pending_address: PendingAddress,
+}
+
+impl Beacon {
+    /// Decode beacon frame from byte  buffer
+    ///
+    /// # Returns
+    ///
+    /// Returns [`Beacon`] and the number of bytes used are returned
+    ///
+    /// # Errors
+    ///
+    /// This function returns an error, if the bytes either don't are enough or
+    /// dont't contain valid data. Please refer to [`DecodeError`] for details.
+    ///
+    /// [`DecodeError`]: ../enum.DecodeError.html
+    /// [`Beacon`]: struct.Beacon.html
+    pub fn decode(buf: &[u8]) -> Result<(Self, usize), DecodeError> {
+        let mut offset = 0usize;
+        let (superframe_spec, size) = SuperframeSpecification::decode(&buf[offset..])?;
+        offset = offset + size;
+        let (guaranteed_time_slot_info, size) =
+            GuaranteedTimeSlotInformation::decode(&buf[offset..])?;
+        offset = offset + size;
+        let (pending_address, size) = PendingAddress::decode(&buf[offset..])?;
+        offset = offset + size;
+        Ok((
+            Beacon {
+                superframe_spec,
+                guaranteed_time_slot_info,
+                pending_address,
+            },
+            offset,
+        ))
+    }
+    /// Encode beacon frame into byte buffer
+    ///
+    /// # Returns
+    ///
+    /// Returns the number of bytes written to the buffer
+    ///
+    /// # Panics
+    ///
+    /// Panics if the buffer is not long enough to hold the frame.
+    ///
+    pub fn encode(&self, buf: &mut [u8]) -> usize {
+        let size = self.superframe_spec.encode(buf);
+        let mut offset = size;
+        let size = self.guaranteed_time_slot_info.encode(&mut buf[offset..]);
+        offset += size;
+        let _ = self.pending_address.encode(&mut buf[offset..]);
```

Comment:
Can you please explain why the second size is ignored, and the first size added twice? Is this intentional?

Reply:
No, this seems like an error. The size from encoding pending address should probably be used,
```suggestion
        let size = self.pending_address.encode(&mut buf[offset..]);
```
I will investigate.